chain/reorg_test.go

```diff
@@ -1,71 +1,141 @@
+	// Add 10 tipsets to the head of the chainStore.
+	requireGrowChain(ctx, t, blockSource, chainStore, 10, dstP)
+	commonAncestor := requireHeadTipset(t, chainStore)
+
+	// make the first fork tipset
+	signer, ki := types.NewMockSignersAndKeyInfo(1)
+	mockSignerPubKey := ki[0].PublicKey()
+	fakeChildParams := th.FakeChildParams{
+		Parent:      commonAncestor,
+		GenesisCid:  dstP.genCid,
+		Signer:      signer,
+		MinerPubKey: mockSignerPubKey,
+		StateRoot:   dstP.genStateRoot,
+		Nonce:       uint64(4),
+	}
+
+	firstForkBlock := th.RequireMkFakeChild(t, fakeChildParams)
+	requirePutBlocks(t, blockSource, firstForkBlock)
+	firstForkTS := th.RequireNewTipSet(t, firstForkBlock)
+	firstForkTsas := &chain.TipSetAndState{
+		TipSet:          firstForkTS,
+		TipSetStateRoot: dstP.genStateRoot,
+	}
+	th.RequirePutTsas(ctx, t, chainStore, firstForkTsas)
+	err := chainStore.SetHead(ctx, firstForkTS)
+	require.NoError(t, err)
+
+	// grow the fork by 10 blocks (11 total)
+	requireGrowChain(ctx, t, blockSource, chainStore, 10, dstP)
+	oldHead := requireHeadTipset(t, chainStore)
+
+	// go back and complete the original chain
+	err = chainStore.SetHead(ctx, commonAncestor)
+	require.NoError(t, err)
+	requireGrowChain(ctx, t, blockSource, chainStore, 14, dstP)
+	newHead := requireHeadTipset(t, chainStore)
+
+	return oldHead, newHead, commonAncestor
+}
+
+func getPrefixOldNewCommon(ctx context.Context, t *testing.T, chainStore chain.Store, blockSource *th.TestFetcher, dstP *DefaultSyncerTestParams) (types.TipSet, types.TipSet, types.TipSet) {
+	// Add 10 tipsets to the head of the chainStore.
+	requireGrowChain(ctx, t, blockSource, chainStore, 10, dstP)
+	oldHead := requireHeadTipset(t, chainStore)
+
+	// Add 2 tipsets to the head
+	requireGrowChain(ctx, t, blockSource, chainStore, 2, dstP)
+	newHead := requireHeadTipset(t, chainStore)
+
+	return oldHead, newHead, oldHead
+}
+
+func getSubsetOldNewCommon(ctx context.Context, t *testing.T, chainStore chain.Store, blockSource *th.TestFetcher, dstP *DefaultSyncerTestParams) (types.TipSet, types.TipSet, types.TipSet) {
+	requireGrowChain(ctx, t, blockSource, chainStore, 10, dstP)
+	commonAncestor := requireHeadTipset(t, chainStore)
+	requireGrowChain(ctx, t, blockSource, chainStore, 1, dstP)
+	oldHead := requireHeadTipset(t, chainStore)
+	headBlock := oldHead.ToSlice()[0]
+
+	signer, ki := types.NewMockSignersAndKeyInfo(1)
+	mockSignerPubKey := ki[0].PublicKey()
+	block2 := th.RequireMkFakeChild(t, th.FakeChildParams{
+		Parent:      commonAncestor,
+		MinerPubKey: mockSignerPubKey,
+		Signer:      signer,
+		GenesisCid:  dstP.genCid,
+		StateRoot:   dstP.genStateRoot})
+	requirePutBlocks(t, blockSource, block2)
+	superset := th.RequireNewTipSet(t, headBlock, block2)
+	supersetTsas := &chain.TipSetAndState{
+		TipSet:          superset,
+		TipSetStateRoot: dstP.genStateRoot,
+	}
+	th.RequirePutTsas(ctx, t, chainStore, supersetTsas)
+
+	return oldHead, superset, commonAncestor
+}
```

Comment:
Move these helper functions to the bottom of the file so that the logic proceeds in a top-down manner.